### PR TITLE
Use a socket to notify controlleragentconfig worker of config reload instead of SIGHUP

### PR DESCRIPTION
**Pull request by Aflynn50**
The current approach of using a SIGHUP signal sent by the controller charm to notify the `controlleragentconfig` worker of a config change only works on machines, not on kubernetes. This PR adds a new socket (`configchange.socket`) that can be used to request a config reload. A `POST` request can be sent to `configchange.socket` at the endpoint `/reload` which will be picked up by the `controlleragentconfig` worker and cause all watchers of the config to be restarted.

This change factors out the socket listener from the `controlsocket` worker to a new package `internal/socketlistener` which is then used in `controlsocket` worker and the `controlleragentconfig` worker.

The `controlsocket` package is refactored to move `handlers.go` and `handlers_test.go` into `worker.go` and `worker_test.go` (which accounts for a decent chunk of the diff of this PR).  

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

### Unit tests
```
go test github.com/juju/juju/internal/worker/controlleragentconfig -check.vv
go test github.com/juju/juju/internal/socketlistener -check.vv
go test github.com/juju/juju/internal/worker/controlsocket -check.vv 
```

### Manual
Manually check that the socket exists and that it responds. This should be checked on both machine and k8s clouds.
```
make microk8s-operator-update
juju bootstrap [cloud] microk8s
juju switch controller
juju ssh controller/0
stat /var/lib/juju/configchange.socket
```
Curl the socket (the socket is protected on machine clouds so `sudo` is needed):
```
(sudo) curl -i -X POST http://localhost/reload  --unix-socket /var/lib/juju/configchange.socket
```
Now check in `juju debug-log -m controller` for:
```
INFO juju.worker.controlleragentconfig reload config request received, reloading config
``` 
### Integration
Get the modified version of the controller charm https://github.com/juju/juju-controller/pull/67
- `charmcraft pack` this charm.
- Bootstrap a new LXD controller using `--controller-charm-path <path-to-packed-artefact>`.
- `juju enable-ha`.
- Once all units have joined the relation, ssh to the controller machine and check _/var/lib/juju/agents/controller-0/agent.conf_. It should be populated along the lines of:
```
db-bind-addresses:
  controller/0: 10.246.27.181
  controller/1: 10.246.27.53
  controller/2: 10.246.27.12
```
- `juju debug-log -m controller --replay|grep "controlleragentconfig reload config request received, reloading config"` indicate the config has reloaded for each controller

## Links

**Jira card:** JUJU-5453

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
Can we add a lot more information to the commits. We want to why the changes are made. More for an archaeological stand point.

File: cmd/jujud-controller/agent/machine/manifolds.go [hunk 1]
```diff
@@ -820,10 +822,11 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
...
 		controlSocketName: ifController(controlsocket.Manifold(controlsocket.ManifoldConfig{
-			StateName:  stateName,
-			Logger:     loggo.GetLogger("juju.worker.controlsocket"),
-			NewWorker:  controlsocket.NewWorker,
-			SocketName: paths.ControlSocket(paths.OSUnixLike),
+			StateName:         stateName,
+			Logger:            loggo.GetLogger("juju.worker.controlsocket"),
+			NewWorker:         controlsocket.NewWorker,
+			NewSocketListener: controlsocket.NewSocketListener,
+			SocketName:        paths.ControlSocket(paths.OSUnixLike),
```
Comment (SimonRichardson) [inline cmd/jujud-controller/agent/machine/manifolds.go:829]: A followup PR is to rename controlSocket to metricSocket.

File: internal/worker/controlleragentconfig/worker.go [hunk 2]
```diff
@@ -107,6 +133,21 @@ func newWorker(cfg WorkerConfig, internalStates chan string) (*configWorker, err
...
+}
+
+// reloadHandler sends a signal to the configWorker when a config reload is
+// requested.
+func (w *configWorker) reloadHandler(resp http.ResponseWriter, req *http.Request) {
+	select {
+	case <-w.catacomb.Dying():
+	case <-req.Context().Done():
+	case w.reloadRequested <- struct{}{}:
+	}
```
Comment (SimonRichardson) [inline internal/worker/controlleragentconfig/worker.go:148]: Let's try and be good citizens and return a good status code.

```suggestion
	select {
	case <-w.catacomb.Dying():
		resp.WriteHeader(http.StatusInternalServerError)
	case <-req.Context().Done():
		resp.WriteHeader(http.StatusInternalServerError)
	case w.reloadRequested <- struct{}{}:
		resp.WriteHeader(http.StatusNoContent)
	}
```

File: internal/worker/controlleragentconfig/worker.go [hunk 3]
```diff
@@ -148,15 +187,21 @@ func (w *configWorker) loop() error {
...
 			return w.catacomb.Err()
-		case <-ch:
+		case <-w.reloadRequested:
 			w.reportInternalState(stateReload)
 
-			w.cfg.Logger.Infof("SIGHUP received, reloading config")
+			// Empty channel of extra reload requests.
+			for len(w.reloadRequested) > 0 {
+				<-w.reloadRequested
+			}
```
Comment (SimonRichardson) [inline internal/worker/controlleragentconfig/worker.go:196]: I think this is the wrong place to do this. Let's just bubble up the changes i.e. remove this.

**Review by SimonRichardson — COMMENTED** [review]
Just a few nips and tucks.

File: internal/worker/controlleragentconfig/worker.go (hunk 3 above)
Comment (Aflynn50) [reply internal/worker/controlleragentconfig/worker.go]: Removed.

File: internal/worker/controlleragentconfig/worker.go (hunk 2 above)
Comment (Aflynn50) [reply internal/worker/controlleragentconfig/worker.go]: Added in the status requests.

**Review by SimonRichardson — APPROVED** [review]
Happy with this change.

It would be good to create a new PR to rename the `controlSocket` to be less generalised and state what it's doing which is metric related things - `metricSocket`

File: internal/worker/controlsocket/worker.go [hunk 4]
```diff
@@ -68,58 +85,252 @@ func NewWorker(config Config) (worker.Worker, error) {
...
+			err := w.config.State.RemoveUser(user.UserTag())
+			if err != nil {
+				// Best we can do here is log an error.
+				w.config.Logger.Warningf("add metrics user failed, but could not clean up user %q: %v",
+					username, err)
+			}
+		}()
+
+	default:
+		return http.StatusInternalServerError, errors.Annotatef(err, "failed to create user %q: %v", username, err)
```
Comment (manadart) [inline internal/worker/controlsocket/worker.go:213]: Annotation is what you were doing, as is the case for the others you've got proximate to here. The fact that it failed is implicit in the error, so: "creating user...".

File: internal/worker/controlleragentconfig/worker.go [hunk 5]
```diff
@@ -107,6 +133,24 @@ func newWorker(cfg WorkerConfig, internalStates chan string) (*configWorker, err
 	return w, nil
 }
 
+func (w *configWorker) registerHandlers(r *mux.Router) {
+	r.HandleFunc("/reload", w.reloadHandler).
+		Methods(http.MethodGet)
```
Comment (manadart) [inline internal/worker/controlleragentconfig/worker.go:138]: Was the choice of GET discussed? I think POST is more intuitive, and might be extended in future such that the body is pertinent...

**Review by manadart — COMMENTED** [review]
Some considerations.

File: internal/socketlistener/socketlistener.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	sl := &SocketListener{
+		config:   config,
+		listener: l,
+	}
+	sl.tomb.Go(sl.run)
+	return sl, nil
+}
+
+// Kill is part of the Worker.worker interface.
+func (sl *SocketListener) Kill() {
```
[the remark this answers is not in the document]
Comment (Aflynn50) [reply internal/socketlistener/socketlistener.go:91]: Fair, I wont do this now but seems reasonable, I guess almost every worker would embed a `TombWorker`.

File: internal/worker/controlsocket/worker.go (hunk 4 above)
Comment (Aflynn50) [reply internal/worker/controlsocket/worker.go:213]: Changed, though this was code moved from `handler.go`, I didn't write it.

File: internal/worker/controlleragentconfig/worker.go (hunk 5 above)
Comment (Aflynn50) [reply internal/worker/controlleragentconfig/worker.go:138]: I don't believe me or Simon had a strong reason to use GET here, I'll change it to POST.

I'll also change it in the PR on the controller charm.

File: internal/socketlistener/socketlistener.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/socketlistener/socketlistener.go:91]: I think there are dragons around potential race conditions, so be aware.

File: internal/worker/controlleragentconfig/worker.go (hunk 5 above)
Comment (Aflynn50) [reply internal/worker/controlleragentconfig/worker.go:138]: Is it worth checking that the body is empty, and throwing an error is it is not?

File: cmd/jujud-controller/agent/machine/manifolds.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2015 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (Aflynn50) [reply cmd/jujud-controller/agent/machine/manifolds.go:1]: Have disabled the Goland feature that does this

File: internal/socketlistener/socketlistener.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/socketlistener/socketlistener.go:91]: Also I think we could reuse this: https://github.com/juju/juju/blob/main/internal/worker/simpleworker.go

File: internal/worker/controlleragentconfig/worker.go (hunk 5 above)
Comment (manadart) [reply internal/worker/controlleragentconfig/worker.go]: Not at this point.

**Review by manadart — APPROVED** [review]
Also tested the metrics socket for regression. All looking good.

Just need to change the package in that _package_test.go_ file.

We want to create a follow-up task for this to inject the user service to the `controlsocket` worker, and use Dqlite to add users instead of Mongo.

I will look at the controller charm next.

**Pull request merged**